blackbird_python/blackbird/tests/test_auxiliary.py

```diff
@@ -0,0 +1,558 @@
+        assert not _literal(nonnumeric)
+
+    def test_literal_bool_invalid(self):
+        """Test that a Blackbird bool is properly converted to a Python type"""
+        nonnumeric = blackbirdParser.NonnumericContext(parser, ctx)
+        nonnumeric.BOOL = lambda: True
+        nonnumeric.getText = lambda: "false"
+
+        with pytest.raises(ValueError, match="Unknown boolean value"):
+            _literal(nonnumeric)
+
+    def test_literal_invalid(self):
+        """Test that a Blackbird bool is properly converted to a Python type"""
+        nonnumeric = blackbirdParser.NonnumericContext(parser, ctx)
+        nonnumeric.getText = lambda: ""
+
+        with pytest.raises(ValueError, match="Unknown value"):
+            _literal(nonnumeric)
+
+
+class TestNumber:
+    """Tests for the _number function"""
+
+    @pytest.mark.parametrize('n, expected', test_ints)
+    def test_number_int(self, parser, ctx, n, expected):
+        """Test that a Blackbird int is properly converted to a Python type"""
+        number = blackbirdParser.NumberContext(parser, ctx)
+        number.INT = lambda: True
+        number.getText = lambda: n
+        assert _number(number) == expected
+
+    @pytest.mark.parametrize('n, expected', test_floats)
+    def test_number_float(self, parser, ctx, n, expected):
+        """Test that a Blackbird float is properly converted to a Python type"""
+        number = blackbirdParser.NumberContext(parser, ctx)
+        number.FLOAT = lambda: True
+        number.getText = lambda: n
+        assert _number(number) == expected
+
+    @pytest.mark.parametrize('n, expected', test_complex)
+    def test_number_complex(self, parser, ctx, n, expected):
+        """Test that a Blackbird complex is properly converted to a Python type"""
+        number = blackbirdParser.NumberContext(parser, ctx)
+        number.COMPLEX = lambda: True
+        number.getText = lambda: n
+        assert _number(number) == expected
+
+    def test_number_pi(self, parser, ctx):
+        """Test that a Blackbird pi literal is properly converted to a Python type"""
+        number = blackbirdParser.NumberContext(parser, ctx)
+        number.PI = lambda: True
+        assert _number(number) == np.pi
+
+    def test_number_invalid(self, parser, ctx):
+        """Test that a Blackbird bool is properly converted to a Python type"""
+        number = blackbirdParser.NumberContext(parser, ctx)
+        number.getText = lambda: ""
+        print(number.PI())
+
+        with pytest.raises(ValueError, match="Unknown number"):
+            _number(number)
+
+
+class TestFunction:
+    """Tests for the _func function"""
+
+    @pytest.mark.parametrize('n, expected', test_complex)
+    def test_function_exp(self, parser, ctx, n, expected, num):
+        """Test that a Blackbird exp function is properly called"""
+        func = blackbirdParser.FunctionContext(parser, ctx)
+        func.EXP = lambda: True
+        expression = num(n)

```

Comment:
I found it hard to understand what `num` function does

Reply:
The `num` pytest fixture creates a node of a 'fake' or dummy abstract syntax tree, that looks like a Blackbird number to the parser. It was created as a fixture since so many tests used it, and this avoids code duplication. I'll update the docstring.